File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int sol = 0;
        int left = 0;
        for(int i = 0; i < s.length() ; i++){
            char c = s.charAt(i);
            int x = map.getOrDefault(c, 0);
            while(x == 1){
                char d = s.charAt(left++);
                if(c == d) x--;
                else map.remove(d);
            }
            map.put(c, 1);
            sol = Math.max(sol, i - left + 1);
        }
        return sol;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = right = 0
        hashset = set()
        longest = 0
        while right < len(s):
            l, r = s[left], s[right]
            if r in hashset:
                hashset.remove(l)
                left += 1
            else:
                hashset.add(r)
                right += 1
                longest = max(longest, right - left)
        return longest
            